projectq/setups/decompositions/h2rx_test.py

```diff
@@ -0,0 +1,119 @@
+#   Copyright 2017 ProjectQ-Framework (www.projectq.ch)
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"Tests for projectq.setups.decompositions.h2rx.py"
+
+import math
+
+import numpy as np
+
+import pytest
+
+from projectq import MainEngine
+from projectq.backends import Simulator
+from projectq.cengines import (AutoReplacer, DecompositionRuleSet,
+                               DummyEngine, InstructionFilter, MainEngine)
+from projectq.meta import Control
+from projectq.ops import Measure, Ph, H, HGate
+
+import h2rx
+
+def test_recognize_correct_gates():
+    """ Test that recognize_HNoCtrl recognizes ctrl qubits """
+    saving_backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=saving_backend)
+    qubit = eng.allocate_qubit()
+    ctrl_qubit = eng.allocate_qubit()
+    eng.flush()
+    H | qubit
+    with Control(eng, ctrl_qubit):
+        H | qubit
+    eng.flush(deallocate_qubits=True)
+    assert h2rx._recognize_HNoCtrl(saving_backend.received_commands[3])
+    assert not h2rx._recognize_HNoCtrl(saving_backend.received_commands[4])
+
+
+def h_decomp_gates(eng, cmd):
+    """ Test that cmd.gate is a gate of class HGate """
+    g = cmd.gate
+    if isinstance(g, HGate): # H is just a shortcut to HGate
+        return False
+    else:
+        return True
+
+# ------------test_decomposition function-------------#
+# Creates two engines, correct_eng and test_eng. 
+# correct_eng implements H gate. 
+# test_eng implements the decomposition of the H gate.
+# correct_qb and test_qb represent results of these two
+# engines, respectively. 
+# The decomposition in this case only produces the 
+# same state as H up to a global phase. 
+# test_vector and correct_vector represent the final 
+# wave states of correct_qb and test_qb. 
+# The dot product of correct_vector and test_vector
+# should have absolute value 1, if the two vectors are the
+# same up to a global phase.
+
+def test_decomposition():
```

Comment:
This test only tests the *_M decomposition rule. Please modify it so that it also covers the *_P decomposition rule.

Something like:
```python
@pytest.mark.parametrize("rule", cnot2rxx.all_defined_decomposition_rules)
def test_decomposition(rule):
    ...
    rule_set = DecompositionRuleSet(rules=[rule])
    ...
```